src/psyclone/psyir/tools/call_tree_utils.py

```diff
@@ -0,0 +1,415 @@
+from psyclone.psyir.tools.read_write_info import ReadWriteInfo
+
+
+# pylint: disable=too-few-public-methods
+class CallTreeUtils():
+    '''This class provides functions functions to analyse the sequence of
+    calls.
+    '''
+
+    # ------------------------------------------------------------------------
+    def _compute_all_non_locals(self, routine):
+        # pylint: disable=too-many-branches
+        '''This function computes all non-local access of the specified
+        routine node. It returns a list of 3-tuples containing:
+        - the type of the non-local as string, one of:
+          - routine: A routine that is called
+          - unknown: an imported symbol, which could still be a function or
+                     an array. This will be resolved later when checking
+                     the module the symbol is imported from.
+          - reference: a variable access
+        - the name of the module the symbol is imported from
+        - the signature of the access
+
+        :param routine: the routine for which to collect all non-local
+            accesses
+        :type routine: :py:class:`psyclone.psyir.nodes.Routine`
+
+        :returns: list of non-local references
+        :rtype: List[Tuple[str, str, :py:class:`psyclone.core.Signature`]]
+
+
+        '''
+        non_locals = []
+
+        outer_module = routine.ancestor(Container)
+
+        for access in routine.walk((Kern, Call, Reference)):
+            # Builtins are certainly not externals, so ignore them.
+            if isinstance(access, BuiltIn):
+                continue
+
+            if isinstance(access, Kern):
+                # A kernel is a subroutine call from a module:
+                non_locals.append(("routine", access.module_name,
+                                   Signature(access.name)))
+                continue
+
+            if isinstance(access, IntrinsicCall):
+                # Intrinsic calls can be ignored
+                continue
+
+            if isinstance(access, Call):
+                sym = access.routine
+                if isinstance(sym.interface, ImportInterface):
+                    module_name = sym.interface.container_symbol.name
+                    non_locals.append(("routine", module_name,
+                                       Signature(sym.name)))
+                    continue
+
+                # No import. This could either be a routine from
+                # this module, or just a global function.
+                if isinstance(sym.interface, DefaultModuleInterface):
+                    # A function defined in the same module
+                    non_locals.append(("routine", outer_module.name,
+                                       Signature(sym.name)))
+                    continue
+
+                # We don't know where the subroutine comes from
+                non_locals.append(("routine", None, Signature(sym.name)))
+                continue
+
+            sym = access.symbol
```

Comment:
Comment please, e.g. "Access must be a Reference"

Reply:
Done.